kvbc/include/categorization/updates.h

```diff
@@ -0,0 +1,150 @@
+#include <string>
+#include <variant>
+
+// Categorized key-value updates for KVBC blocks. Every category supports different properties and functionalities.
+// Note1: Empty category IDs are invalid and not supported.
+// Note2: Using the same category ID for different category types is an error.
+namespace concord::kvbc::categorization {
+
+// Shared updates across multiple categories.
+// Persists key-values directly in the underlying key-value store. All key-values are marked stale during the update
+// itself. Explicit deletes are not supported.
+struct SharedKeyValueUpdates {
+  SharedKeyValueUpdates() {}
+  SharedKeyValueUpdates(SharedKeyValueUpdates&& other) = default;
+
+  struct SharedValue {
+    SharedValue(std::string&& val, std::set<std::string>&& cat_ids) {
+      data_.value = std::move(val);
+      for (auto it = cat_ids.begin(); it != cat_ids.end();) {
+        data_.categories_ids.emplace_back(cat_ids.extract(it++).value());
+      }
+    }
+
+   private:
+    SharedValueData data_;
+    friend class SharedKeyValueUpdates;
+  };
+
+  void addUpdate(std::string&& key, SharedValue&& val) { data_.kv.emplace(std::move(key), std::move(val.data_)); }
+
+ private:
+  SharedUpdatesData data_;
+};
+
+// Updates for a key-value category.
+// Persists key-values directly in the underlying key-value store.
+// Controls whether a hash of the updated key-values is calculated for this update
+struct KeyValueUpdates {
+  KeyValueUpdates() {}
+
+  KeyValueUpdates(KeyValueUpdates&& other) = default;
+  struct Value {
+    std::string data;
+
+    // If set, the expiry time will be persisted and available on lookups.
+    // An implementation might optionally choose to automatically mark the key-value stale at or after the `expire_at`
+    // time. If not done automatically, the application needs to manage the lifetime of the key.
+    std::optional<std::time_t> expire_at;
+
+    // Mark the key-value stale during the update itself.
+    bool stale_on_update{false};
+  };
+
+  void addUpdate(std::string&& key, Value&& val) {
+    data_.kv.emplace(std::move(key), ValueWithExpirationData{std::move(val.data), val.expire_at, val.stale_on_update});
+  }
+
+  // Set a value with no expiration and that is not stale on update
+  void addUpdate(std::string&& key, std::string&& val) {
+    data_.kv.emplace(std::move(key), ValueWithExpirationData{std::move(val), 0, false});
+  }
+
+  void addDelete(std::string&& key) {
+    if (const auto [itr, inserted] = unique_deletes_.insert(key); !inserted) {
+      // already inserted, error?
+      return;
+    }
+    data_.deletes.emplace_back(std::move(key));
+  }
+
+  void calculateHash(const bool hash) { data_.calculate_hash = hash; }
+
+ private:
+  KeyValueUpdatesData data_;
+  std::set<std::string> unique_deletes_;
+};
+
+// Updates for a merkle tree category.
+// Persists key-values in a merkle tree that is constructed on top of the underlying key-value store.
+struct MerkleUpdates {
+  MerkleUpdates() {}
```

Comment:
As discussed offline, we can maybe make the default ctor, move ctor and move assignment `=default` and the copy ctor and copy assignment `=delete` in order to be explicit.